Keeper retries during insert

devcrafter opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support for keeper request retries during insert into replicated merge trees. Apart from fault tolerance, it aims to provide better user experience, - avoid returning a user an error during insert if keeper is restarted (for example, due to upgrade).

It also introduces failure injection mode for keeper requests during insert into replicated merge tree, so it's possible to inject a failure with certain probability when accessing keeper interface (see `insert_keeper_fault_injection_probability`). It's also possible to reproduce a particular insert execution with failure injections by providing the same seed, see `insert_keeper_fault_injection_seed`

Closes https://github.com/ClickHouse/ClickHouse/issues/39764

devcrafter committed:
Tries control struct

devcrafter committed:
commit parts with retries (no retires for quorum yet)

devcrafter committed:
Fix style

+ do not count try if there is no error

devcrafter committed:
Fix style check

qoega commented:
Let's test it doing rolling restart of cluster of keepers in integration test as well? We should not fail inserts in this case?

devcrafter committed:
Retries for quorum + some fixes

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Do not retry in case of logical errors (not zk connectivity)

devcrafter committed:
Fixes

devcrafter committed:
Fix quorum retries

devcrafter committed:
More natural flow control

+ no break is necessary in retry loop
+ zk -> keeper renaming

devcrafter committed:
Merge branch 'master' into igor/insert_zk_retries

devcrafter committed:
Merge branch 'master' into igor/insert_zk_retries

alexey-milovidov commented:
Changelog entry is required.

devcrafter committed:
Separation of retires and their control logic

+ integration test (not working yet)

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

Commit:
Automatic style fix

devcrafter committed:
+ retries for ReplicatedMergeTreeSink::consume execution
+ intergration tests with disconnect and keeper restart

devcrafter committed:
StorageReplicatedMergeTree::write shouldn't check read only mode since
there are retries

Commit:
Automatic style fix

devcrafter committed:
Use the same retries_info for consume() execution path

devcrafter committed:
Fix style check

devcrafter committed:
Fix style

devcrafter committed:
draft: KeeperAccess class for fault injections

devcrafter committed:
Compilable version

devcrafter committed:
Fault injection: something working

+ logging

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Fix typos

devcrafter committed:
+ retry loop method
+ print seed on KeeperAccess destruction, so we can reproduce the run
+ change settings

devcrafter committed:
No fault injections by default

devcrafter committed:
Fix: setting description

devcrafter committed:
Fix: crash + keeper calls diagnostic

devcrafter committed:
sql test for new settings

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Fix: missing assignment to replicas number

devcrafter committed:
Use KeeperAccess for StorageReplicatedMergeTree::allocateBlockNumber()

devcrafter committed:
Fix zookeeper utils

devcrafter committed:
Revert "Fix zookeeper utils"

This reverts commit 3d6a73ce9737fecdbd7ea14bdd08b310174cbc1b.

devcrafter committed:
Remove unnecessary change

devcrafter committed:
Improved dianostic

devcrafter committed:
Update test: add SYNC to drop queries

devcrafter committed:
Fix typo

devcrafter committed:
Randomize settings for injection failures

devcrafter committed:
Some polishing

devcrafter committed:
Increase number of retries

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -579,42 +621,62 @@ void ReplicatedMergeTreeSink::commitPart(
-                throw Exception("Unexpected logical error while adding block " + toString(block_number) + " with ID '" + block_id + "': "
-                                + Coordination::errorMessage(multi_code) + ", path " + failed_op_path,
-                                ErrorCodes::UNEXPECTED_ZOOKEEPER_ERROR);
+                throw Exception(
+                    ErrorCodes::UNEXPECTED_ZOOKEEPER_ERROR,
+                    "Unexpected logical error while adding block {} with ID '{}': {}, path {}",
```
Do I understand correctly, that we will not retry on this error? (It's probably ok, because the error message looks frightening and probably it's not retryable, just want to clarify)

tavplubix replied:
Maybe we should try to replace error code with `LOGICAL_ERROR` (so it will crash in debug build)

devcrafter replied:
Here I just reformat the message. It's not a connectivity error with zookeeper, so I don't really want to touch it here

devcrafter commented on src/Storages/MergeTree/KeeperAccess.h:
```diff
@@ -0,0 +1,236 @@
+namespace DB
+{
+
+class FaultInjection
```
Yes, I had 2 implementations initially. But finally decided to keep only one for now because another implementation (`FailEachCallOnce`) would require `KeeperAccess` and retries controllers to work together. Which I decided to not do for now.

devcrafter commented on src/Core/Settings.h:
```diff
@@ -657,6 +657,12 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
+    M(UInt64, insert_keeper_max_retries, 9, "Max retries for keeper operations during insert", 0) \
+    M(UInt64, insert_keeper_retry_initial_backoff_ms, 100, "Initial backoff timeout for keeper operations during insert", 0) \
+    M(UInt64, insert_keeper_retry_max_backoff_ms, 10000, "Max backoff timeout for keeper operations during insert", 0) \
+    M(UInt64, insert_keeper_fault_injection_mode, 0, "0 - off, 1 - random (see insert_keeper_fault_injection_seed)", 0) \
+    M(UInt64, insert_keeper_fault_injection_seed, 0, "0 - random seed, otherwise the setting value", 0) \
+    M(Float, insert_keeper_fault_injection_probability, 0.1f, "Approximate probability of failure for a keeper request during insert. Valid value is in interval [0.0f, 1.0f]", 0) \
```
Originally, it's here because of [this](https://github.com/ClickHouse/ClickHouse/pull/41734#discussion_r986006568). We can remove it for now, and introduce it later if we'd decide to implement yet another mode

devcrafter commented on src/Storages/MergeTree/KeeperAccess.h:
```diff
@@ -0,0 +1,236 @@
+    std::bernoulli_distribution distribution;
+};
+
+class KeeperAccess
```
I see. `ZooKeeperWrapper` and `ZooKeeperAdapter` looks too generic, `ZooKeeperWithFaultInjection` a bit long but would work

devcrafter commented on src/Storages/MergeTree/EphemeralLockInZooKeeper.h:
```diff
@@ -86,14 +88,14 @@ class EphemeralLockInZooKeeper : public boost::noncopyable
 
 std::optional<EphemeralLockInZooKeeper> createEphemeralLockInZooKeeper(
-    const String & path_prefix_, const String & temp_path, zkutil::ZooKeeper & zookeeper_, const String & deduplication_path);
+    const String & path_prefix_, const String & temp_path, const KeeperAccessPtr & zookeeper_, const String & deduplication_path);
```
`ZooKeeper` client is not created on insert execution path. So, we have no control over its creation, → just using it.

tavplubix replied:
Hm, yes, but we can wrap real `zkutil::ZooKeeper` client into an object that implements the same interface and calls methods of wrapped client. We will not need to replace `zkutil::ZooKeeper` with `KeeperAccessPtr` everywhere. However, it will require to make all methods virtual (and make an interface) and wrap all of them (a lot of boilerplate). So I'm not sure which approach is better, maybe `KeeperAccessPtr` is okay.

tavplubix replied:
Another (questionable and quite dirty) approach: introduce a thread_local pointer to a struct with fault injection settings (or put it into `ThreadStatus`). Create an object that stores these settings, sets the pointer in ctor and resets it in dtor. Check the pointer in `ZooKeeper::pushRequest` (before `try` to avoid finalization) and throw exception if needed. 

But I do not insist on any these approaches, feel free to keep `KeeperAccessPtr`.

devcrafter commented on src/Storages/MergeTree/KeeperRetries.h:
```diff
@@ -0,0 +1,214 @@
+            return false;
+        }
+
+        /// the flag will set to false in case of error
+        iteration_succeeded = true;
```
We just need to set it to `true` for each new iteration. I guess you prefer to do it after retry limit check, - will do

devcrafter committed:
Slight polishing

devcrafter committed:
Merge branch 'igor/insert_zk_retries_polishing' into igor/insert_zk_retries

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter commented on src/Storages/MergeTree/KeeperAccess.h:
```diff
@@ -0,0 +1,236 @@
+public:
+    using Ptr = std::shared_ptr<KeeperAccess>;
+
+    static KeeperAccess::Ptr createInstance(
```
That is because `make_shared` requires public constructor. But in case of fault injection, constructing object in less efficient way will not hurt, so ok, will address it

devcrafter committed:
Addressing review comments

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -91,34 +91,45 @@ static void assertSessionIsNotExpired(zkutil::ZooKeeperPtr & zookeeper)
     size_t replicas_number = replicas.size();
     size_t quorum_size = getQuorumSize(replicas_number);
 
     if (active_replicas < quorum_size)
+    {
+        if (Coordination::isHardwareError(keeper_error))
+            throw Coordination::Exception("Failed to check number of alive replicas", keeper_error);
```
I guess, moving initiation of 'is current replica active' request ([line 108](https://github.com/ClickHouse/ClickHouse/pull/41734/files#diff-d34dfcef2ee825d0184c61973c0c3c8606d68975870f22c9ddd1501f7ff88e42R108)) before `replicas_status_futures` initialization loop will break this assumption. I'd rather keep this keeper error check here, to be independent of the requests order

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -455,7 +487,13 @@ void ReplicatedMergeTreeSink::commitPart(
+                                waitForQuorum(
+                                    zookeeper, existing_part_name, quorum_path, quorum_info.is_active_node_value, replicas_num);
+                            }))
+                        return;
```
No, in case of exception in `waitForQuorum`, the exception is caught in `call()`, set corresponding error in retries control structure and `call()` returns false. Then execution will return from lambda to retry loop (see `retryLoop()`). There we'll see that there was an error, check if we can retry one more time, if so, then will retry

devcrafter committed:
Remove deleted setting from test randomized settings

devcrafter committed:
Attempt to fix: `Unexpected state of part ... (state Outdated). Expected: TemporaryFix`

devcrafter committed:
Rename KeeperAccess -> ZooKeeperWithFaultInjection

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Rename KeeperRetries -> ZooKeeperRetries

devcrafter committed:
Handle more accurately zk errors during unlocking ephemeral lock

+ KeeperRetries -> ZooKeeperRetries

devcrafter committed:
Handle ZNONDE case during unlocking epehemeral lock

devcrafter committed:
Finalize keeper session in case of fault injection

https://pastila.nl/?019d8cb1/130aae1741526ba9bcf1d3564a45a3ad

devcrafter committed:
Merge branch 'master' into igor/insert_zk_retries

devcrafter committed:
Reduce probablity of failure injection

... since we close(finalize) zookeeper client in case of failure injection

+ additional diagnostic

devcrafter committed:
Return session expired for all calls after first fault injection during
retry/iteration

- reseting zk session for real is too harsh, it affect another test
  executed in parallel
+ increasing faulure injection probability to 0.1 since no more global
  zk client reset

devcrafter committed:
Check on retry if part info was successfully written into ZK on previous
try but we got ZK error due to network issue

devcrafter committed:
Fix typos

devcrafter committed:
Fix error: unexpected state for part during retry

Unexpected state of <part_name> (state Active). Expected: Temporary . (NOT_FOUND_EXPECTED_DATA_PART)

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Try to fix case where part is written on disk with one name but in zk
with another

https://pastila.nl/?019d8cb1/1c6273f3e97acd830e33d2dcf3d9fa42

devcrafter committed:
Best effort retries

if part is renamed but failed to write into zk due to connectivity error,
check on retry if we successfully wrote into into zk, - if not - fail
insert

devcrafter committed:
Fix: uninitialized action_after_last_failed_retry

devcrafter committed:
Inject failure for tryMultiNoThrow only after operation

devcrafter committed:
Fix: fault injection cleanup

+ delete ephemeral node in case of fault injection after successful
  ephemeral node creation

devcrafter committed:
Fix style

devcrafter committed:
Adjust keeper retries timeout for inserts to retry faster during test

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Fault injection cleanup for multi requests

devcrafter committed:
Fix style and spelling

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Adjust insert_keeper* setting for test

01586_replicated_mutations_empty_partition

devcrafter committed:
Disable fault injection in tests where we check part ids

+ disable 01459_manual_write_to_replicas.sh for now, there is a bug with
  deduplication

devcrafter committed:
Adjust insert retries setting for tests

devcrafter committed:
Disable fault injection in 02369_lost_part_intersecting_merges

devcrafter committed:
Adjust test for keeper fault injection

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Adjust test which can be non-deterministic due to keeper fault injections on inserts

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Fix build: adopt accessor to ZooKeeper multi exists/get methods

devcrafter committed:
Align class names and logs

devcrafter committed:
Rename header according to class names

devcrafter commented on src/Storages/MergeTree/KeeperRetries.h:
```diff
@@ -0,0 +1,214 @@
+            throw Exception(user_error.code, user_error.message);
+
+        if (keeper_error.code != KeeperError::Code::ZOK)
+            throw zkutil::KeeperException(keeper_error.code, keeper_error.message);
```
Nice consideration, I thought about it. But since a source error is not always triggered via exception, I simply handled it by storing the error code and message as a common denominator. We can improve it later.

tavplubix commented on tests/queries/0_stateless/00121_drop_column_zookeeper.sql:
```diff
@@ -1,14 +1,14 @@
 -- Tag no-replicated-database: Old syntax is not allowed
 
-DROP TABLE IF EXISTS alter_00121;
+DROP TABLE IF EXISTS alter_00121 SYNC;
```
It does not really make sense to edit all tests and add `SYNC` modifier to every `DROP` query, you can simply enable `database_atomic_wait_for_drop_and_detach_synchronously` in you config (as it's enabled in CI for convenience)

devcrafter replied:
Thanks for mentioning it. I modified only tests I run locally. Setting `database_atomic_wait_for_drop_and_detach_synchronously` is a good approach, but in general, it's nice when a stuff just works.

devcrafter committed:
Fault injection in lockSharedData/unlockSharedData

- not complete

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Some polishing

Algunenano commented:
Some notes about usability:

* IMO It'd be much more comfortable to have a setting for the max wait time for ZK. The current proposal with 3 (`insert_keeper_max_retries`, `insert_keeper_retry_initial_backoff_ms` and `insert_keeper_retry_max_backoff_ms`) forces you to make calculations . This is important because `max_execution_time` isn't being considered in the sleep/wait, so it might be that you wait 40 seconds for ZK to recover just to get cancelled a moment later by having reached the query timeout (30 seconds ago).

* A nice addition to the previous for visibility would be to add a profile event to see how much time was spent waiting for keeper. Right now you can see there were issues `ZooKeeperTransactions':3,'ZooKeeperGet':1,'ZooKeeperMulti':2,'ZooKeeperHardwareExceptions':10` but it's hard to know how impactful that was.

In any case, this new feature is great and helps to reduce complexity on the user side.

devcrafter committed:
Forgot to provide zookeeper pointer to lockSharedData/unlockSharedData

devcrafter committed:
Complete fault injection in lockSharedData/unlockSharedData

tavplubix commented on src/Storages/MergeTree/ZooKeeperRetries.h:
```diff
@@ -0,0 +1,250 @@
+            {
+                f();
+            }
+            catch (const zkutil::KeeperException & e)
+            {
+                if (!Coordination::isHardwareError(e.code))
+                    throw;
+
+                setKeeperError(e.code, e.message());
+            }
```
It will not retry on `TABLE_IS_READ_ONLY` (`StorageReplicatedMergeTree::write` may throw this)

tavplubix replied:
From `StorageReplicatedMergeTree::write`

devcrafter committed:
Merge branch 'master' into igor/insert_zk_retries

tavplubix commented on src/Storages/MergeTree/ZooKeeperRetries.h:
```diff
@@ -0,0 +1,250 @@
+
+    Coordination::Error getLastKeeperErrorCode() const { return keeper_error.code; }
+
+    void actionAfterLastFailedRetry(std::function<void()> f) { action_after_last_failed_retry = std::move(f); }
```
Should we save (and call) all lambdas that were ever passed to this function? Should we call them if the last retry succeeded? Or this function can be called only once? Imagine that we are unlucky and on each iteration we get zk error on attempt to commit part to zk. The lambda passed to this function captures `part_name`. If we try to insert part with new name on next iteration, then we will forget to call `enqueuePartForCheck` for previously inserted parts. Seems like it's not possible, because after `actionAfterLastFailedRetry` we will either return or throw in the beginning of the next iteration (and will never try to allocate new block number), but it's not clear from the code and the interface. Consider adding some assertions or moving code that checks if part was committed to zk to the branch that handles connection loss on commit.

devcrafter replied:
Added comments in #42607 [c53b96a](https://github.com/ClickHouse/ClickHouse/pull/42607/commits/c53b96a11ce776f96e7cbd3b8cfe74f9fd495526)

devcrafter committed:
Overload allocateBlockNumber to provide version for fault injections

devcrafter committed:
Fix: more correct behavior of try*() method in fault injection impl

try*() methods return error code but in case of connectivity errors
throw
this can be improved since different try*() method can returns error
(not throws) on different set of error, i.e. some can throw on
non-connectivity errors as well (tryGet() for example)

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Fix tryMultiNoThrow() wrapper

devcrafter committed:
Fix typo

devcrafter commented:
todo:
- [ ] parallel quorum timeout, `01509_parallel_quorum_and_merge_long`, https://s3.amazonaws.com/clickhouse-test-reports/41734/2203ef88e1d2a4f74cf27de1e2b244f997cc2fc7/stateless_tests__msan__[1/3].html

devcrafter committed:
Handle zk exception for lockSharedData()

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Disable fault injections by default

+ enable fault injection for functional tests

devcrafter committed:
Clean up ephemeral nodes after each iteration in retryLoop

When keeper failure is injected, session expiration is emulated.
It should lead to ephemeral nodes removal. So, we do it too.

+ addressing some review comments

devcrafter committed:
Fix style

devcrafter committed:
Disable keeper fault injections on insert for couple tests

devcrafter committed:
Remove debug logging

devcrafter committed:
Patch from @antonio2368, makes fault injection implementation more concise

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter committed:
Simplify integration test

Commit:
Automatic style fix

devcrafter committed:
Fix crash when no paths provided for multi paths requests

+ fix some review comments

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries

devcrafter commented:
Further changes will be done ([here](https://github.com/ClickHouse/ClickHouse/pull/42607), please check description for the reason behind)

devcrafter commented on src/Storages/MergeTree/EphemeralLockInZooKeeper.cpp:
```diff
@@ -41,11 +49,15 @@ std::optional<EphemeralLockInZooKeeper> createEphemeralLockInZooKeeper(
         if (e != Coordination::Error::ZOK)
         {
-            if (responses[0]->error == Coordination::Error::ZNODEEXISTS)
+            if (!responses.empty() && responses.front()->error == Coordination::Error::ZNODEEXISTS)
```
Fixed in https://github.com/ClickHouse/ClickHouse/pull/42607 [a601c16](https://github.com/ClickHouse/ClickHouse/pull/42607/commits/a601c166ad43e96366cb975a0d862725d9a6dd69)
